ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/participant/state/kvutils/api/BatchingLedgerWriterSpec.scala

```diff
@@ -0,0 +1,226 @@
+        res2 should be(SubmissionResult.Acknowledged)
+        batches should be(Seq(Seq(subm2), Seq(subm1)))
+        queue.alive should be(true)
+      }
+    }
+
+    "commit batch after max batch size exceeded" in {
+      var batches = Seq.empty[Seq[DamlSubmissionBatch.CorrelatedSubmission]]
+      val queue =
+        DefaultBatchingQueue(10, 15, 50.millis, 1)
+          .run { batch =>
+            {
+              batches = batch +: batches
+              Future.successful(())
+            }
+          }
+
+      val subm1 = DamlSubmissionBatch.CorrelatedSubmission.newBuilder
+        .setCorrelationId("1")
+        .setSubmission(ByteString.copyFromUtf8("helloworld"))
+        .build
+      val subm2 = DamlSubmissionBatch.CorrelatedSubmission.newBuilder
+        .setCorrelationId("2")
+        .setSubmission(ByteString.copyFromUtf8("helloworld"))
+        .build
+
+      for {
+        res1 <- queue.offer(subm1)
+        _ = {
+          batches.size should be(0)
+        }
+        res2 <- queue.offer(subm2)
+      } yield {
+        // First batch emitted.
+        batches.size should be(1)
+
+        // Wait for second batch.
+        eventually {
+          batches.size should be(2)
+        }
+
+        res1 should be(SubmissionResult.Acknowledged)
+        res2 should be(SubmissionResult.Acknowledged)
+        batches should be(Seq(Seq(subm2), Seq(subm1)))
+        queue.alive should be(true)
+      }
+    }
+  }
+
+  def immediateBatchingQueue: BatchingQueue =
+    new BatchingQueue {
+      override def run(commitBatch: Seq[DamlSubmissionBatch.CorrelatedSubmission] => Future[Unit])(
+          implicit materializer: Materializer): BatchingQueueHandle =
+        new BatchingQueueHandle {
+          override def alive: Boolean = true
+          override def offer(
+              submission: DamlSubmissionBatch.CorrelatedSubmission): Future[SubmissionResult] =
+            commitBatch(Seq(submission))
+              .map { _ =>
+                SubmissionResult.Acknowledged
+              }
+        }
+    }
+
+  "BatchingLedgerWriter" should {
+
+    "construct batch correctly" in {
+      val batchCaptor = MockitoHelpers.captor[kvutils.Bytes]
+      val mockWriter = createWriter(Some(batchCaptor))
+      val batchingWriter =
+        LoggingContext.newLoggingContext { implicit logCtx =>
+          new BatchingLedgerWriter(immediateBatchingQueue, mockWriter)
+        }
+      val (corId1, subm1) = "test1" -> ByteString.copyFrom(Array[Byte](1, 2, 3))
+      val expected = createExpectedBatch(corId1, subm1)
+      for {
+        res1 <- batchingWriter.commit(corId1, subm1)
+      } yield {
+        verify(mockWriter).commit(anyString(), ArgumentMatchers.eq(expected))
+        res1 should be(SubmissionResult.Acknowledged)
+      }
+    }
+
+    "drop batch when commit fails" in {
+      val mockWriter = createWriter(None, SubmissionResult.Overloaded)
+      val batchingWriter = createBatchingWriter(mockWriter, 5, 1.millis)
```

Comment:
Here readability could be improved by introducing `aSubmission` (as recommended above) and setting `maxBatchSizeBytes = aSubmission.size` explicitly.

Reply:
agreed!